eclair-core/src/main/scala/fr/acinq/eclair/wire/internal/channel/version0/ChannelCodecs0.scala

```diff
@@ -143,19 +145,33 @@ private[channel] object ChannelCodecs0 {
       })
     )
 
-    val htlcTxAndSigsCodec: Codec[HtlcTxAndSigs] = (
-      ("txinfo" | txWithInputInfoCodec.downcast[HtlcTx]) ::
-        ("localSig" | variableSizeBytes(uint16, sig64OrDERCodec)) :: // we store as variable length for historical purposes (we used to store as DER encoded)
-        ("remoteSig" | variableSizeBytes(uint16, sig64OrDERCodec))).as[HtlcTxAndSigs].decodeOnly
-
-    val publishableTxsCodec: Codec[PublishableTxs] = (
-      ("commitTx" | (("inputInfo" | inputInfoCodec) :: ("tx" | txCodec)).as[CommitTx]) ::
-        ("htlcTxsAndSigs" | listOfN(uint16, htlcTxAndSigsCodec))).as[PublishableTxs].decodeOnly
-
-    val localCommitCodec: Codec[LocalCommit] = (
+    val htlcTxsAndRemoteSigsCodec: Codec[HtlcTxAndRemoteSig] = (
```

Comment:
I took a different approach in my migration, which I think is more future-proof:

1. Keep `HtlcTxAndSigs` and the previous `Commitments` object in `ChannelTypes0`
2. Provide a `migrate` function on that type that returns the new `Commitments` class with the right data in it
3. Use it in `ChannelCodecs0`, `ChannelCodecs1` and `ChannelCodecs2` (similar to what was done for the `XxxCommitPublished` [here](https://github.com/ACINQ/eclair/blob/45204e238052ca1691f9f6486eaf1863ae8836b2/eclair-core/src/main/scala/fr/acinq/eclair/wire/internal/channel/version0/ChannelCodecs0.scala#L245))

It makes the changes to the `ChannelCodecsN` files minimal (just add a call to `.migrate()`) and puts the logic in `ChannelTypes0` which makes the `migrate()` function very easy to unit test (instead of having to provide blobs of old serialized data).

I think this is the only way we'll be able to really scale changes to the channel classes with full backwards-compatibility. If for example we decide in version 12 of eclair to change again the format of the `Commitments` case class, we'd simply have to "snapshot" the current `Commitments` in `ChannelTypes3.scala` (and it would be a different snapshot than the one in `ChannelTypes0.scala`). In the `ChannelTypes0.Commitments.migrate()` function we would only need to append a second `.migrate()` call (that would call `ChannelTypes3.Commitments.migrate()`) instead of duplicating some code. WDYT?

Reply:
As much as I like hacking around with `scodec`, this sounds like a more reasonable approach indeed, will give it a try.

There is another unsolved issue by the way: for this migration we probably want to actively migrate all data, as opposed to doing it next time the channel state is updated. I'm not sure what is the best way to do that.

Reply:
> for this migration we probably want to actively migrate all data

In that case we should probably add a DB migration, that simply reads then writes without changing the SQL schemas (or maybe we can also make schema changes if there are pending ones that we postponed?)

Reply:
Done with e4a0ae9. It looks pretty good, but I'm not sure if it would make an additional migration on the same class easier.